### PR TITLE
Merge asm implementations

### PR DESCRIPTION
**Pull request by jonas-schievink**
This replaces the implementation of `inline-asm` with the file I wrote in #259 (and some fixes).

All functions that call assembly now do so via a `call_asm!` macro that either dispatches to a call to an `#[inline(always)]` function containing the inline `asm!`, or to the FFI shim. This makes all functions that call into asm significantly shorter.

The FFI shim is now also macro-generated, which makes it very small.

### CODE REVIEW

**Review by adamgreig — APPROVED** [review]
This looks great, nice work! I'll hold off on the merge to give everyone a chance to have a look over it. Once it's merged let's sort out the remaining fixmes/todos and I'd like to modify the methods that use explicit registers to instead use generic named registers that the compiler can allocate.

**Review by therealprof — APPROVED** [review]
Nice!

File: asm/lib.rs
Comment (jonas-schievink) [inline asm/lib.rs:76]:
```suggestion
// v7m *AND* v8m.main, but *NOT* v8m.base 
#[cfg(any(armv7m, armv8m_main))]
```

File: src/asm.rs
Comment (Rahix) [inline src/asm.rs:24]: I recently dug this up for unrelated reasons.  I just wanted to mention, this match construct was introduced because of rust-embedded/cortex-m#63 with the actual change being from rust-embedded/cortex-m#71.  Is it no longer relevant?

Comment (jonas-schievink) [reply src/asm.rs:24]: It still builds fine for me. Maybe not with `inline-asm`. I should be able to add this feature back though.

Comment (jonas-schievink) [reply src/asm.rs:24]: Ah, I remember now, I removed this to avoid all the warnings this generates. These were the reason almost the entire crate used function arguments like `_val`. I think we should leave this as-is for now.

**Review by therealprof — APPROVED** [review]

**Review by therealprof — CHANGES_REQUESTED** [review]
Actually it looks like the MSRV has changed again, yet README.md still says 1.31

**Review by therealprof — APPROVED** [review]
Thanks!

bors r+

**Pull request merged**
